EWPP-3633: Add support for CKEditor 5

brummbar opened the pull request.

brummbar committed:
EWPP-3633: Port the zebra striping plugin.

brummbar committed:
EWPP-3633: Port the table simple plugin.

brummbar committed:
EWPP-3633: Add icons for simple and zebra plugins.

brummbar committed:
EWPP-3633: Port the table sort plugin.

brummbar committed:
EWPP-3633: Clean sortable attribute when table header rows are removed.

brummbar committed:
EWPP-3633: Add test coverage for the simple and zebra-striped plugins.

brummbar committed:
EWPP-3633: Add test coverage for the column sort plugin.

brummbar committed:
EWPP-3633: Add test coverage for plugin supported elements and multiple header rows.

brummbar committed:
EWPP-3633: Prevent overlapping of dropdown menu on the table.

brummbar committed:
EWPP-3633: Update column sort icon.

brummbar committed:
EWPP-3633: Remove unneeded else().

brummbar committed:
EWPP-3633: Neatify code.

sergepavle commented on modules/oe_theme_helper/tests/src/FunctionalJavascript/Ckeditor5TableTest.php:
```diff
@@ -0,0 +1,435 @@
+      'weight' => 1,
+      'roles' => [RoleInterface::AUTHENTICATED_ID],
+      'filters' => [
+        'filter_ecl_table' => [
```
Should we add test coverage to assert that plugins are available if enabled **filter_ecl_table** and table is available in the toolbar?

brummbar replied:
Absolutely.

sergepavle commented on modules/oe_theme_helper/oe_theme_helper.ckeditor5.yml:
```diff
@@ -0,0 +1,50 @@
+    conditions:
+      plugins:
+        - ckeditor5_table
+      filter: filter_ecl_table
```
For some reason, the condition doesn't work. It looks like conditions work with OR logic.

brummbar replied:
I found out that you actually need a specific order of conditions to apply them correctly.
Whenever a "plugins" condition is present, the check bails out and doesn't evaluate the other condition types.
This because the plugins can be evaluated only when everything else has been evaluated.
But using the correct order (moving plugins always below) it will apply them correctly.

brummbar committed:
EWPP-3633: Add test coverage for plugins conditions.

brummbar committed:
EWPP-3633: Plugins conditions should always be listed last.

nagyad commented on modules/oe_theme_helper/webpack.config.js:
```diff
@@ -0,0 +1,66 @@
+module.exports = [];
+// Loop through every subdirectory in src, each a different plugin, and build
+// each one in ./build.
+getDirectories('./js/ckeditor5_plugins').forEach((dir) => {
```
Shouldn't we do this in a try-catch if readdirSync throws an error?

nagyad commented on modules/oe_theme_helper/webpack.config.js:
```diff
@@ -0,0 +1,66 @@
+      moduleIds: 'named',
+    },
+    entry: {
+      path: path.resolve(
```
I think `path:` is not needed there, since webpack entry takes the path as a string or an array of strings. You can just simply add path.resolve(....

nagyad commented on modules/oe_theme_helper/webpack.config.js:
```diff
@@ -0,0 +1,66 @@
+const { styles, builds } = require('@ckeditor/ckeditor5-dev-utils');
+const TerserPlugin = require('terser-webpack-plugin');
+
+function getDirectories(srcpath) {
```
This function should have an annotation describing its purpose.

brummbar replied:
This file is the same as the webpack.config.js present in core, with modifications provided by the ckeditor5_dev module (maintained by core people). I would not introduce changes to it, for ease of updates in the future and consistency. This is valid for the points above too

brummbar replied:
For reference: https://git.drupalcode.org/project/ckeditor5_dev/-/blob/1.0.x/ckeditor5_plugin_starter_template/webpack.config.js

brummbar committed:
EWPP-3633: Do not use dynamic properties.

brummbar committed:
EWPP-3633: Wait on the visible status of the iframe.

The previouse code waited for an iframe with a hidden status, which was causing
ten seconds of wait wile the real element we need is the visible iframe.

brummbar committed:
EWPP-3633: Lock behat/mink-selenium2-driver to prevent issues with key up/down events.

brummbar committed:
EWPP-3633: Temporarely lock triple-store version.

It will be fixed in a dedicated issue.

brummbar committed:
EWPP-3633: Wait for link to be clickable.
